File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.py
import heapq

n = int(input())
m = int(input())
tree = [[] for _ in range(n + 1)]

for _ in range(m):
    a, b, w = map(int, input().split())
    tree[a].append([w, b])
    tree[b].append([w, a])

hq = []
visited = [False] * (n + 1)
for i in tree[1]:
    heapq.heappush(hq, (i[0], i[1]))
visited[1] = True

result = 0
while hq:
    w, b = heapq.heappop(hq)
    if not visited[b]:
        visited[b] = True
        result += w
        for i in tree[b]:
            heapq.heappush(hq, (i[0], i[1]))

print(result)

File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.java
import java.io.*;
import java.util.*;

public class Main {
    
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int n, m, result;
    static List<Edge>[] tree;
    static boolean[] visited;
    static PriorityQueue<Edge> edges;
    
    static class Edge {
        
        int vertex;
        int weight;
        
        public Edge(int vertex, int weight) {
            this.vertex = vertex;
            this.weight = weight;
        }
    }
    
    static void input() throws IOException {
        n = Integer.parseInt(br.readLine());
        m = Integer.parseInt(br.readLine());
        tree = new List[n + 1];
        for (int i = 1; i < n + 1; i++) {
            tree[i] = new ArrayList<>();
        }
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int vertex1 = Integer.parseInt(st.nextToken());
            int vertex2 = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());
            tree[vertex1].add(new Edge(vertex2, weight));
            tree[vertex2].add(new Edge(vertex1, weight));
        }
    }
    
    static void process() throws IOException {
        result = 0;
        edges = new PriorityQueue<>((e1, e2) -> e1.weight - e2.weight);
        visited = new boolean[n + 1];
        
        edges.addAll(tree[1]);
        visited[1] = true;
        
        while (!edges.isEmpty()) {
            Edge curEdge = edges.poll();
            if (visited[curEdge.vertex]) {
                continue;
            }
            visited[curEdge.vertex] = true;
            result += curEdge.weight;
            
            edges.addAll(tree[curEdge.vertex]);
        }
    }
    
    static void output() throws IOException {
        bw.write(result + "\n");
        bw.flush();
        bw.close();
    }
    
    public static void main(String[] args) throws IOException {
        input();
        process();
        output();
    }
}